### PR TITLE
*: support select from partition table

### PR DESCRIPTION
**Pull request by tiancaiamao**
This PR is ready for review.

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -353,6 +361,38 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 	return bs
 }
 
+// locatePartition returns the partition ID of the input record.
+func (t *Table) locatePartition(ctx sessionctx.Context, r []types.Datum) (int64, error) {
```
Comment (coocood) [inline table/tables/tables.go]: We should store the expression in `*Table` in `TableFromMeta`.

**Comment by tiancaiamao** [comment]
Because I thought it's more difficult to do that, so I choose to modify logical plan.
Anyway, I agree that your idea is more natural. @winoros

**Comment by tiancaiamao** [comment]
This PR is ready for review.
I'll add config and some time in another PR, as this one is big enough. 
PTAL @coocood @winoros

**Comment by tiancaiamao** [comment]
PTAL @coocood @winoros

**Comment by tiancaiamao** [comment]
Done. @coocood

File: plan/rule_table_partition.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	PartitionExpr() []expression.Expression
+}
+
+func prunePartition(sel *LogicalSelection, ds *DataSource) (LogicalPlan, error) {
+	pi := ds.tableInfo.GetPartitionInfo()
+	if pi == nil {
+		return selectOnSomething(sel, ds)
+	}
+
+	var partitionExpr []expression.Expression
```
Comment (coocood) [inline plan/rule_table_partition.go]: partitionExprs

File: plan/rule_table_partition.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	if sel != nil {
+		sel.SetChildren(lp)
+		return sel, nil
+	}
+	return lp, nil
+}
+
+// partitionTable is for those tables which implement partition.
+type partitionTable interface {
+	PartitionExpr() []expression.Expression
```
Comment (coocood) [inline plan/rule_table_partition.go]: PartitionExprs

File: plan/rule_table_partition.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+	// Rewrite data source to union all partitions, during which we may prune some
+	// partitions according to selection condition.
+	children := make([]LogicalPlan, 0, len(pi.Definitions))
+	var selConds []expression.Expression
+	if sel != nil {
+		selConds = sel.Conditions
+	}
+
+	col := partitionExprAccessColumn(partitionExpr[0])
```
Comment (coocood) [inline plan/rule_table_partition.go]: check nil here instead of in `canBePrune`?

**Comment by tiancaiamao** [comment]
PTAL @coocood @winoros

**Comment by tiancaiamao** [comment]
PTAL @coocood @winoros @zz-jason

File: plan/rule_table_partition.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+//      p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// tablePartition is here because it's easier to prune partition after predicate push down.
+type tablePartition struct{}
```
Comment (zz-jason) [inline plan/rule_table_partition.go]: s/tablePartition/partitionPrunning/?

File: plan/rule_table_partition.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+				return nil, errors.Trace(err)
+			}
+			children[i] = child1
+		}
+	}
+
+	return selectOnSomething(sel, lp)
+}
+
+func selectOnSomething(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
```
Comment (zz-jason) [inline plan/rule_table_partition.go]: I think we can make all the functions this file to be a member function of `tablePartition`

File: plan/rule_table_partition.go (hunk 5 above)
Comment (tiancaiamao) [reply plan/rule_table_partition.go]: It not only does the partition pruning, but also the ast structure rewrite, so I think partitionPrunning is not very accurate

**Comment by tiancaiamao** [comment]
PTAL @zz-jason @winoros @coocood

File: plan/rule_table_partition.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (winoros) [inline plan/rule_table_partition.go]: 2018

File: plan/rule_table_partition.go (hunk 5 above)
Comment (coocood) [reply plan/rule_table_partition.go]: s/tablePartition/partitionPrunner

File: plan/rule_table_partition.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+	return s.rewriteDataSource(nil, lp)
+}
+
+func (s *tablePartition) rewriteDataSource(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
+	// Assert there will not be sel -> sel in the ast.
+	switch lp.(type) {
+	case *DataSource:
+		return s.prunePartition(sel, lp.(*DataSource))
+	case *LogicalSelection:
+		return s.rewriteDataSource(lp.(*LogicalSelection), lp.Children()[0])
```
Comment (coocood) [inline plan/rule_table_partition.go]: The argument `sel` is discarded if the plan is Selection.

**Comment by coocood** [comment]
We can simplify the prunner to only consider push down conditions.

File: plan/logical_plan_builder.go [hunk 9]
```diff
@@ -1709,6 +1709,10 @@ func (b *planBuilder) buildDataSource(tn *ast.TableName) LogicalPlan {
 	tableInfo := tbl.Meta()
 	b.visitInfo = appendVisitInfo(b.visitInfo, mysql.SelectPriv, dbName.L, tableInfo.Name.L, "")
 
+	if tableInfo.GetPartitionInfo() != nil {
+		b.optFlag = b.optFlag | flagTablePartition
```
Comment (shenli) [inline plan/logical_plan_builder.go]: We should clear the flag before leaving buildDataSource.

File: plan/rule_table_partition.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+				return nil, errors.Trace(err)
+			}
+			children[i] = child1
+		}
+	}
+
+	return s.selectOnSomething(sel, lp)
+}
+
+func (s *tablePartition) selectOnSomething(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
```
Comment (shenli) [inline plan/rule_table_partition.go]: The function name is weird.

File: plan/rule_partition_processor.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func (s *partitionProcessor) prunePartition(ds *DataSource) (LogicalPlan, error) {
+	pi := ds.tableInfo.GetPartitionInfo()
+	if pi == nil {
+		return ds, nil
+	}
+
+	var partitionExprs []expression.Expression
+	if itf, ok := ds.table.(partitionTable); ok {
```
Comment (coocood) [inline plan/rule_partition_processor.go]: what does `itf` mean?

File: plan/logical_plan_builder.go (hunk 9 above)
Comment (tiancaiamao) [reply plan/logical_plan_builder.go]: We new a planBuilder instance everytime in the `Optimize` function, the object would not be reused, so the flag need not to be clear.

File: plan/rule_table_partition.go (hunk 10 above)
Comment (tiancaiamao) [reply plan/rule_table_partition.go]: Removed.

File: plan/rule_partition_processor.go (hunk 11 above)
Comment (tiancaiamao) [reply plan/rule_partition_processor.go]: interface.
Any suggestion?

File: plan/rule_partition_processor.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+func (s *partitionProcessor) rewriteDataSource(lp LogicalPlan) (LogicalPlan, error) {
+	// Assert there will not be sel -> sel in the ast.
+	switch lp.(type) {
+	case *DataSource:
+		return s.prunePartition(lp.(*DataSource))
+	default:
+		children := lp.Children()
+		for i, child := range children {
+			child1, err := s.rewriteDataSource(child)
```
Comment (zz-jason) [inline plan/rule_partition_processor.go]: s/child1/newChild/

File: plan/rule_partition_processor.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+//      p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// partitionProcessor is here because it's easier to prune partition after predicate push down.
+type partitionProcessor struct{}
```
Comment (shenli) [inline plan/rule_partition_processor.go]: partitionPrunner is a better name.

File: plan/rule_partition_processor.go [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+	if pi == nil {
+		return ds, nil
+	}
+
+	var partitionExprs []expression.Expression
+	if itf, ok := ds.table.(partitionTable); ok {
+		partitionExprs = itf.PartitionExprCache().PartitionPrune
+	}
+	if len(partitionExprs) == 0 {
+		return nil, errors.New("partition expression missing")
```
Comment (zz-jason) [inline plan/rule_partition_processor.go]: how about define an error in `plan/errors.go`?

**Comment by coocood** [comment]
LGTM

File: plan/rule_partition_processor.go (hunk 11 above)
Comment (shenli) [reply plan/rule_partition_processor.go]: pt

File: plan/rule_partition_processor.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+	return lp, nil
+}
+
+// partitionTable is for those tables which implement partition.
+type partitionTable interface {
+	PartitionExprCache() *tables.PartitionExprCache
+}
+
+func (s *partitionProcessor) prunePartition(ds *DataSource) (LogicalPlan, error) {
```
Comment (shenli) [inline plan/rule_partition_processor.go]: s/prunePartition/prune/

File: plan/rule_partition_processor.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+	var partitionExprs []expression.Expression
+	if table, ok := ds.table.(partitionTable); ok {
+		partitionExprs = table.PartitionExprCache().PartitionPrune
+	}
+	if len(partitionExprs) == 0 {
+		return nil, errors.New("partition expression missing")
+	}
+
+	// Rewrite data source to union all partitions, during which we may prune some
+	// partitions according to selection condition.
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply plan/rule_partition_processor.go]: Yeah? 'according to the filter conditions which are pushed to the DataSource'?

**Comment by tiancaiamao** [comment]
PTAL  @zz-jason @shenli

File: plan/rule_partition_prunner.go [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+//    partition p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// partitionPrunner is here because it's easier to prune partition after predicate push down.
+type partitionPrunner struct{}
```
Comment (coocood) [inline plan/rule_partition_prunner.go]: It does more than partition pruning, including rewriting data source.
`partitionProcessor` is more appropriate.

Comment (tiancaiamao) [reply plan/rule_partition_prunner.go]: So, what do you think? @shenli @zz-jason

File: plan/rule_partition_processor.go [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+// partitionProcessor rewrites the ast for table partition.
+//
+// create table t (id int) partition by range (id)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply plan/rule_partition_processor.go]: Fixed

File: plan/rule_partition_processor.go (hunk 13 above)
Comment (tiancaiamao) [reply plan/rule_partition_processor.go]: @coocood  suggest to to use `partitionProcessor`

**Comment by tiancaiamao** [comment]
PTAL @zz-jason @shenli

**Comment by tiancaiamao** [comment]
PTAL @zz-jason

File: table/tables/tables.go [hunk 19]
```diff
@@ -60,8 +60,23 @@ type Table struct {
...
+// Given a compare function, for any partition range i, if cmp[i] > 0, then cmp[i+1] > 0.
+// While partition prune must use the accurate range to do prunning.
+// partition by range (x)
+//   (partition
+//      p1 values less than (y1)
+//      p2 values less than (y2)
+//      p3 values less than (y3))
+// PartitionPrune: (x < y1); (y1 <= x < y2); (y2 <= x < y3)
+// Locate: (x < y1); (x < y2); (x < y3)
+type PartitionExprCache struct {
```
Comment (shenli) [inline table/tables/tables.go]: s/PartitionExprCache/PartitionExpr/

File: table/tables/tables.go [hunk 20]
```diff
@@ -407,18 +448,19 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 // locatePartition returns the partition ID of the input record.
 func (t *Table) locatePartition(ctx sessionctx.Context, pi *model.PartitionInfo, r []types.Datum) (int64, error) {
 	// TODO: Remove this later, when we have better way to TestPartitionAddRecord.
-	if t.partitionExpr == nil {
-		partitionExpr, err := generatePartitionExpr(t.meta)
+	if t.partitionExprCache == nil {
```
Comment (shenli) [inline table/tables/tables.go]: Is there concurrency issue?

File: plan/rule_partition_processor.go [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+	// Calculate the column range to prune.
+	accessConds := ranger.ExtractAccessConditionsForColumn(conds, col.ColName)
+	r, err := ranger.BuildColumnRange(accessConds, ctx.GetSessionVars().StmtCtx, col.RetType)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return len(r) == 0, nil
+}
+
+// partitionExprAccessColumn extracts the column visited from the partition expression.
```
Comment (shenli) [inline plan/rule_partition_processor.go]: Can not understand this comment. 

> the column visited from the partition expression

File: table/tables/tables.go (hunk 20 above)
Comment (tiancaiamao) [reply table/tables/tables.go]: Read and write will not happen at the same time.
This line will just be used in test, notice the TODO.

File: plan/rule_partition_processor.go (hunk 21 above)
Comment (tiancaiamao) [reply plan/rule_partition_processor.go]: s/from/by

**Comment by tiancaiamao** [comment]
PTAL @shenli

**Comment by tiancaiamao** [comment]
PTAL @shenli

File: plan/rule_partition_processor.go [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+	// Calculate the column range to prune.
+	accessConds := ranger.ExtractAccessConditionsForColumn(conds, col.ColName)
+	r, err := ranger.BuildColumnRange(accessConds, ctx.GetSessionVars().StmtCtx, col.RetType)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	return len(r) == 0, nil
+}
+
+// partitionExprAccessColumn extracts the column which is visited by the partition expression.
```
Comment (shenli) [inline plan/rule_partition_processor.go]: s/visited/used/

**Comment by tiancaiamao** [comment]
PTAL @shenli

**Comment by tiancaiamao** [comment]
PTAL @shenlii

**Comment by shenli** [comment]
LGTM

**Pull request merged**